### PR TITLE
Feature/validate plugin options

### PR DESCRIPTION
**Pull request by gribnoysup**
**What kind of change does this PR introduce?**

A feature

**Did you add tests for your changes?**

Not yet, I feel that I need some feedback first

**If relevant, link to documentation update:**

Not yet, but maybe needed?

**Summary**

A possible solution for https://github.com/webpack/webpack/issues/364

This PR addresses problem with validation of webpack built-in plugins options. Sometimes there are some checks in place, sometimes there are none. Some option patterns for plugins are confusing, this change could help to identify configuration problems quicker

This PR adds json-schemas and validation for almost all documented build-in webpack plugins from https://webpack.js.org/plugins/:

- AggressiveSplittingPlugin
- CommonsChunkPlugin
- LimitChunkCountPlugin
- MinChunkSizePlugin
- BannerPlugin
- DllPlugin
- DllReferencePlugin
- HashedModuleIdsPlugin
- LoaderOptionsPlugin
- SourceMapDevToolPlugin
- WatchIgnorePlugin

Changes made:

- new `WebpackError` created: `WebpackPluginValidationError`
- common methods from `WebpackOptionsValidationError` that are shared with `WebpackPluginValidationError` moved to `/util`
- added `validateSchema` to plugins

**Does this PR introduce a breaking change?**

Possibly?

**Other information**

The one thing that bothers me the most is that I found some undocumented options while running the tests. This makes me unsure that the schemas I made by the docs are valid and/or fully covers plugins options.

Another thing is I'm not sure about is how to cover (or should they be covered at all?) plugins like [ContextReplacementPlugin](https://webpack.js.org/plugins/context-replacement-plugin/) with this approach, as they don't follow the more common "one `options` argument" approach in other plugins

I was also thinking about sharing some common `definitions` between webpack/plugin options schemas, but this will mean changing `json` files to `js` probably

Am I going with this in the right direction in general?

### CODE REVIEW

File: lib/BannerPlugin.js [hunk 1]
```diff
@@ -8,6 +8,10 @@
 const ConcatSource = require("webpack-sources").ConcatSource;
 const ModuleFilenameHelpers = require("./ModuleFilenameHelpers");
 
+const validateSchema = require("./validateSchema");
```
Comment (michael-ciniawsky) [inline lib/BannerPlugin.js]: @gribnoysup You could use [`schema-utils`](https://github.com/webpack-contrib/schema-utils) for that purpose as standardize way for plugins/loaders. If you need anything feel free to open an issue in `webpack-contrib/schema-utils`

```js
const schema = require("../schemas/plugins/bannerPluginSchema.json");
const validateOptions = require("schema-utils");

class Plugin {
  constructor (options) {
     validateOptions(schema, options, `${name}` /* <= Banner Plugin */)
  }
}
```

### `ValidationError`

validationerror

File: lib/util/indent.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+"use strict";
```
Comment (michael-ciniawsky) [inline lib/util/indent.js]: This file shouldn't be necessary with `schema-utils`

File: lib/util/getSchemaPart.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+"use strict";
```
Comment (michael-ciniawsky) [inline lib/util/getSchemaPart.js]: This file shouldn't be necessary with `schema-utils`

File: lib/util/formatSchema.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+"use strict";
```
Comment (michael-ciniawsky) [inline lib/util/formatSchema.js]: This file shouldn't be necessary with `schema-utils`

File: lib/util/getSchemaPartText.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+"use strict";
```
Comment (michael-ciniawsky) [inline lib/util/getSchemaPartText.js]: This file shouldn't be necessary with `schema-utils`

File: lib/BannerPlugin.js (hunk 1 above)
Comment (gribnoysup) [reply lib/BannerPlugin.js]: @michael-ciniawsky almost done with making requested changes, but I think I'm blocked by https://github.com/webpack-contrib/schema-utils/issues/3 at the moment 😿

I made a PR with the possible solution!

Comment (michael-ciniawsky) [reply lib/BannerPlugin.js]: I will try to get your PR in as soon as possible 😛  thx again

**Comment by gribnoysup** [comment]
A little update: 

Blocking PR to the `schema-utils` was merged, but there are some problems with NPM and releasing rights (https://github.com/webpack-contrib/schema-utils/pull/10#issuecomment-340226735). I tested locally with the new `schema-utils` and everything works just fine, just waiting for the new version to be actually released.

**Comment by gribnoysup** [comment]
Seems like TravisCi experiencing some issues right now

image

File: package.json [hunk 6]
```diff
@@ -116,6 +117,7 @@
     "cover:unit": "node --max-old-space-size=4096 --harmony --trace-deprecation ./node_modules/istanbul/lib/cli.js cover --report none node_modules/mocha/bin/_mocha -- test/*.unittest.js",
     "cover:report": "istanbul report",
     "cover:report-min": "istanbul report --report lcovonly",
-    "publish-patch": "npm run lint && npm run beautify-lint && mocha && npm version patch && git push && git push --tags && npm publish"
+    "publish-patch": "npm run lint && npm run beautify-lint && mocha && npm version patch && git push && git push --tags && npm publish",
+    "webpack-defaults": "webpack-defaults"
```
Comment (michael-ciniawsky) [inline package.json]: Why is this needed ?  That's the boilerplate for repos we use at `webpack-contrib` which doesn't apply here :)

Comment (gribnoysup) [reply package.json]: @michael-ciniawsky Sorry! Didn't notice this change at first. I removed it in the latest commit, this is no longer in the branch.

When I was testing with the `schema-utils` installed from the git tag version, I had to build it manually inside of `node_modules` and somehow build step added this line to `package.json`, still don't really understand why this happened 😅

**Comment by gribnoysup** [comment]
@michael-ciniawsky 

Is there any way to trigger TravisCI again without pushing? Are there any more changes I need to make?

**Comment by michael-ciniawsky** [comment]
> Is there any way to trigger TravisCI again without pushing?

I couldn't restart the build, gives an Error 😥 . Try

```
git commit --amend --no-edit && git push <repo> <branch> -f
```

>  Are there any more changes I need to make?

No, I don't think so. The `${name}` could with a space `DLLPlugin` => `DLL Plugin` but that's entirely preference and a total nitpick 🐦

**Comment by gribnoysup** [comment]
@michael-ciniawsky that is actually a good idea, error message will be more aligned with Extract Text Plugin that already uses `schema-utils` to validate options and this also will trigger travis ci again 😸

File: lib/SourceMapDevToolPlugin.js [hunk 7]
```diff
@@ -53,12 +56,9 @@ class SourceMapDevToolPlugin {
...
 		if(arguments.length > 1)
 			throw new Error("SourceMapDevToolPlugin only takes one argument (pass an options object)");
-		// TODO: remove in webpack 3
-		if(typeof options === "string") {
-			options = {
-				sourceMapFilename: options
-			};
-		}
+
+		validateOptions(sourceMapDevToolPluginSchema, options || {}, "Source Map Dev Tool Plugin");
```
Comment (michael-ciniawsky) [inline lib/SourceMapDevToolPlugin.js]: `SourceMap DevTool Plugin` ?

**Comment by michael-ciniawsky** [comment]
@gribnoysup Can you make one a of the plugins fail and post the `console.log`. I tried a Test Plugin a few hours ago and got a weird output, we may need to change something in `schema-utils` in regard to plugins

**Comment by gribnoysup** [comment]
@michael-ciniawsky it's just printing `false` on the validation error and the same issue is with ExtractTextPlugin too 😱  Seems like `this.stack = false;` in `schema-utils` is causing this.

I should probably add some tests to validate plugin error messages actually showing something before this can be merged

**Comment by michael-ciniawsky** [comment]
Yep, you're right using `Error.captureStackTrace(this. this.constructor)` fixes it half way :) but it's still off somehow and doesn't seem to get highlighted at all

```
<root>/__test__/node_modules/schema-utils/dist/validateOptions.js:40 // ?
    throw new _ValidationError2.default(ajv.errors, name);           // ?
    ^

ValidationError:

Extract Text Plugin (Loader) Invalid Options

options['foobar'] should NOT have additional properties
```

**Comment by gribnoysup** [comment]
@michael-ciniawsky I think that the highlighting for the `WebpackOptionsValidationError` comes from this part of code 

https://github.com/webpack/webpack/blob/master/bin/webpack.js#L337 

and is not applied to any other errors. And this part is also the part that ignores `stack` when printing the error. I may be wrong here, but sure seems like so, are there any other cases where webpack highlights compilation errors by itself?

Maybe we should add the check for ValidationError also?

```
if(e instanceof WebpackOptionsValidationError || e instanceof ValidationError) {
```

**Comment by michael-ciniawsky** [comment]
#5913

**Comment by gribnoysup** [comment]
@michael-ciniawsky no, please, go ahead! I'm happy I could help 🤗

>1. 'Scoped' const schema per file ?
>e.g const bannerPluginSchema => const schema

Yep, I should totally change that!

>2. Already in ./schema/plugins, so Schema in the filename kind of redundant ?
>e.g "../schemas/plugins/bannerPluginSchema.json" => "../schemas/plugins/BannerPlugin.json"

I like this, but I made those names based on how webpackOptionsSchema was named, I think it could be renamed too, maybe? There is really seems to be no need in including `Schema` in the name of the files in this folder

**Comment by michael-ciniawsky** [comment]
@gribnoysup Could you try with `schema-utils >= v0.4.2` it should be working fine now :)

File: schemas/plugins/optimize/aggressiveSplittingPluginSchema.json [hunk 8]
```diff
@@ -0,0 +1,23 @@
+{
+	"id": "AggressiveSplittingPlugin",
```
Comment (michael-ciniawsky) [inline schemas/plugins/optimize/aggressiveSplittingPluginSchema.json]: Please use two spaces/one tab for indent

Comment (michael-ciniawsky) [reply schemas/plugins/optimize/aggressiveSplittingPluginSchema.json]: If this is just github displaying tabs `\t` weirdly forget what I said then 😛

Comment (gribnoysup) [reply schemas/plugins/optimize/aggressiveSplittingPluginSchema.json]: I think this is just a github issue 😸 

image

**Comment by gribnoysup** [comment]
@michael-ciniawsky I had problems while trying to add tests because of `schema-utils@0.4.2` calling `process.exit(1)` on the error, finishing tests process completely. Are the tests needed for this case or having test coverage at `schema-utils` library is enough?

Everything else is fixed and updated! 

I also wanted to make error messages more helpful and aligned with the current webpack error messaging (because default error messages are not very good), but this will be another PR for `schema-utils` 😜

**Comment by michael-ciniawsky** [comment]
Terminating the process via `process.exit(1)` is necessary to have consistent behaviour and `err.messages` for loaders && plugins (if there is a better way to handle it, feel free to tell me all about it :stuck_out_tongue:). Where did you try to add tests ? `test/binCases` ? I test it in `schema-utils` via an `ENV` variable which doesn't terminate the process when set. If you have better mesages in mind please open a PR over at `schema-utils` :+1:

**Comment by gribnoysup** [comment]
@michael-ciniawsky oh, I missed the `binCases`, seems just like the place for this test, thanks!

File: test/binCases/errors/plugin-options-validation/test.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+"use strict";
+
+module.exports = function testAssertions(code, stdout, stderr) {
+	code.should.be.eql(1);
+	stdout.should.be.empty();
+
+	stderr[1].should.containEql("Commons Chunk Plugin Invalid Options");
+
+	stderr[3].should.containEql("options should be string");
```
Comment (michael-ciniawsky) [inline test/binCases/errors/plugin-options-validation/test.js]: Did you mean this messages by 'not good ' ? 😛 These are `defaults` ? Do we need them ? Based on the `webpack.config.js` I only expected `"options['foo'] should NOT have additional properties"` here, but that's up to sokra to decide

Comment (gribnoysup) [reply test/binCases/errors/plugin-options-validation/test.js]: This and the actual error messaging format is not very good. In this case, `oneOf` is the reason for these error messages: CommonsChunkPlugin options could be a string, array or object and AJV reports errors on every `oneOf` as if all of them failed. 

This happens in the webpackOptionValidation too, but the [error messaging there is changed](https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsValidationError.js#L144) in a way, that will print this error a lit more nicely. It will look something like this:

```
options has an unknown property 'foo'

options should be one of these:
* string
* array[string]
* { <object schema> }
```

**Comment by gribnoysup** [comment]
Rebased on master and moved tests. Also had to add support for such nested tests to `BinTestCases` 😸

File: test/BinTestCases.test.js [hunk 10]
```diff
@@ -41,115 +41,114 @@ function convertToArrayOfLines(outputArray) {
...
-								}
-
-								const stdout = convertToArrayOfLines(env.stdout);
-								const stderr = convertToArrayOfLines(env.stderr);
-								testAssertions(stdout, stderr, done);
-								child.kill();
-							}, 8000); // wait a little to get an output
-						});
+		if(asyncExists) {
+			describe(testName, function() {
```
Comment (michael-ciniawsky) [inline test/BinTestCases.test.js]: I think all `function()` could be refactored to an arrow, but that's entirely additional chore so feel free to ignore 😛

File: test/binCases/errors/validation/plugins/test.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+"use strict";
+
+module.exports = function testAssertions(code, stdout, stderr) {
```
Comment (michael-ciniawsky) [inline test/binCases/errors/validation/plugins/test.js]:
```diff
- function testAssertions(code, stdout, stderr) 
+ function(code, stdout, stderr) 
```

File: test/binCases/errors/plugin-options-validation/test.js (hunk 9 above)
Comment (michael-ciniawsky) [reply test/binCases/errors/plugin-options-validation/test.js]: We can definitely handle `anyOf, oneOf, default` etc in a better way with better messages, `schema-utils` does the minimum atm. If you have something in mind feel free to open an issue or PR if you have the time.

File: test/binCases/errors/validation/plugins/webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+var path = require("path");
```
Comment (michael-ciniawsky) [inline test/binCases/errors/validation/plugins/webpack.config.js]: `var` => `const`

**Comment by gribnoysup** [comment]
I first converted everything in tests to arrow functions, but context of `mocha` is used in some of them, and arrow functions break this, also using arrow functions [is actually discouraged](http://mochajs.org/#arrow-functions) by `mocha` team. So for consistency I left `it` and `describe` functions as it is, only changing all other ones

**Comment by gribnoysup** [comment]
Fixed conflicts with master

**Comment by michael-ciniawsky** [comment]
> This is not good. It should throw instead. Never call process.exit

I added this recently to `schema-utils` (mainlyfor this PR) behave the same/align with the way [webpack does here](https://github.com/webpack/webpack/blob/master/bin/webpack.js#L345) bc I didn't found a better way to handle plugins validation errors. Is there a better way to do it ? Without exiting the process the `err.message` for plugins will be out of whack and for loaders it will be displayed multiple times

**Comment by sokra** [comment]
Basically a simple guideline is to not use `console` and `process.exit`, except in the CLI. The CLI should be the only one handling user interaction. It's weird to get `console` logs when using the node.js API and it's super weird to see the process exiting. Even more when you don't get an stack trace, like in the test result above:

> ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․webpack: Using compiler.parser is deprecated.
> Use compiler.plugin("compilation", function(compilation, data) {
>   data.normalModuleFactory.plugin("parser", function(parser, options) { parser.apply(/* ... */); });
> }); instead. It was called at Context.<anonymous> (/home/travis/build/webpack/webpack/test/Compiler.test.js:184:22).
> ․․․․․․․․․․․,,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
> Dll Plugin Invalid Options
> 
> options.name is a required property

No idea where this is coming from.

The solution is to handle validation error in a pretty way in the CLI, like how validation errors from the webpack configuration are handled: https://github.com/webpack/webpack/blob/ad9b957474a1c03e146ce68169154f4cf7e4dcf2/bin/webpack.js#L336-L349

Technically it's also bad to call `process.exit` in this case (and we should refactor this). When calling `process.exit` it's not guaranteed that pending console messages are printed to the console. So in this case it's not guaranteed that the error message is printed.

It's better to do it this way: https://github.com/webpack/webpack/blob/8915dbbde61df11ec30a4623614f27b34f0f9905/bin/webpack.js#L155-L159

**Comment by gribnoysup** [comment]
@sokra @michael-ciniawsky as I understand there are two changes needed: we need to remove `process.exit` from `schema-utils` and then change webpack bin to handle errors more gracefully. Am I right?

**Comment by sokra** [comment]
yep

**Comment by michael-ciniawsky** [comment]
Yep, I will take a look into this as soon as possible, how should the graceful handle look like? The problem in `schema-utils` are throwing  plugins which aren't 'gracefully' handled by webpack atm. Is displaying the `err.message` and 'rethrowing' an `{Error}` instead of `process.exit(1)`  sufficient ? But that's a suggestion out of the blue, might be easier and better handled :).  I also think I tried that, but can't remember why I didn't went that route tbh 😛 . The messages needed some improvement aswell as @gribnoysup  suggested (via [ajv-errors](https://github.com/epoberezkin/ajv-errors) maybe or cherry-picked from webpack).

**Comment by michael-ciniawsky** [comment]
@sokar Displaying a (limited) `Error.stackTrace` is prefered/better in your opinion ?

**Comment by sokra** [comment]
> Is displaying the err.message and 'rethrowing' an {Error} instead of process.exit(1) sufficient

Display an error message and set `process.exitCode = 1` than return from CLI. (If not easily possible, call process.exit like in the other validation message code.)

> The messages need some improvement aswell as @gribnoysup suggested (via ajv-errors maybe or cherry-picked from webpack).

I've added a lot of extra code to create useful and readable error messages (and process nested errors) here: https://github.com/webpack/webpack/blob/46a6c349fbf267862dd7c144dd48a88ced00105e/lib/WebpackOptionsValidationError.js#L74-L216

It would be great to move this to schema-utils to have better messages for plugins too. webpack could use schema-utils for the options validation too. I guess this would be better in a separate PR.

**Comment by sokra** [comment]
> Displaying a (limited) Error.stackTrace is prefered/better in your opinion ?

It would actually be great to display a stacktrace starting from webpack.config.js.

This can be done by wrapping the configuration evaluation in a custom marker function `(function WEBPACK_CONFIG() { ... })()` when calling it and remove all messages after the marker function in the stack trace. This is done for loaders too:

https://github.com/webpack/loader-runner/blob/0d41655430a52af88e6a423c6d8df5c22d413e81/lib/LoaderRunner.js#L118-L120

``` js
		var result = (function LOADER_EXECUTION() {
			return fn.apply(context, args);
		}());
```

https://github.com/webpack/webpack/blob/46a6c349fbf267862dd7c144dd48a88ced00105e/lib/ErrorHelpers.js#L7-L15

**Comment by gribnoysup** [comment]
> This can be done by wrapping the configuration evaluation in a custom marker function (function WEBPACK_CONFIG() { ... })() when calling it and remove all messages after the marker function in the stack trace

@sokra @michael-ciniawsky can I jump on this feature in this branch also? Or maybe in another PR? The solution suggested by @sokra is very interesting and I want to try to implement it!

**Comment by michael-ciniawsky** [comment]
@gribnoysup If you have a solution in mind simply propose it I would say 😛 , if it is better to open a separate PR then you can always split it out later then :)

**Comment by michael-ciniawsky** [comment]
Could someone please point me to where the config evaluation happens ? @gribnoysup Are you working on something already (no rush etc, just want to know if there is any progress made already, I could also contrib to your/a fork if needed then) ?

**Comment by gribnoysup** [comment]
@michael-ciniawsky sorry for the delayed response! I'm a little stuck with this too. As I understand, here is where evaluation happens:

https://github.com/webpack/webpack/blob/2525466d6be52b4a096a125e15d5a49f62e0ad34/bin/convert-argv.js#L96-L100

I'm actually not sure that there is any way to actually point to `webpack.config.js` file on webpack options validation without somehow saving the path and adding it to the stack output manually. As I understand, config evaluation happens in a different stack that config validation, therefore there is no trace of calling it when the error is thrown. So wrapping evaluation to a method as in loader case won't help. So yeah, I don't have any other ideas how to do this

**Comment by sokra** [comment]
This PR is about adding validation to the plugin constructors (not the webpack configuration validation). The constructor calls do happen inside of `requireConfig`.

**Comment by gribnoysup** [comment]
@sokra oh, I misunderstood what you meant, sorry!

**Comment by gribnoysup** [comment]
Okay, then its pretty easy to catch that trace! But first, we need to change `schema-utils` to not exit process and just throw an error. 

The only thing I'm worried about is that the module trace is pretty big and it is right between our marker and webpack config. Here is an example trace from error, thrown by the plugin when `requireConfig` is wrapped in marker:

```
    at validateOptions (/Users/s.petushkov/Projects/webpack/node_modules/schema-utils/dist/validateOptions.js:56:13)
    at new CommonsChunkPlugin (/Users/s.petushkov/Projects/webpack/lib/optimize/CommonsChunkPlugin.js:31:3)
    at Object.<anonymous> (/Users/s.petushkov/Projects/webpack/examples/chunkhash/webpack.config.js:14:3)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at WEBPACK_OPTIONS (/Users/s.petushkov/Projects/webpack/bin/convert-argv.js:97:55)
    at requireConfig (/Users/s.petushkov/Projects/webpack/bin/convert-argv.js:97:78)
    at /Users/s.petushkov/Projects/webpack/bin/convert-argv.js:104:17

    // ... more traces
```

@sokra @michael-ciniawsky sorry for bothering you again! Do you think I can clean up trace first by `WEBPACK_OPTIONS` and then by `module.js`? Or maybe just throw away everything after the line with `webpack.config.js` in it?

**Comment by michael-ciniawsky** [comment]
> Okay, then its pretty easy to catch that trace! But first, we need to change schema-utils to not exit process and just throw an error.

@gribnoysup I will do this asap as soon as we have a rough draft of how it will work, you can meanwhile use the ugly 'hack' I used for testing `JEST=true webpack` so `schema-utils` will an throw `{Error}` instead

Could you show the wrapper code please ? 

> Do you think I can clean up trace first by WEBPACK_OPTIONS and then by module.js? Or maybe just throw away everything after the line with webpack.config.js in it?

If there isn't anything that makes this buggy under certain circumstances (different `stack` in env x), the  `err.stackTrace` should be limited at least imho. Maybe it's even possible to 'parse' the stack and format the `err.message` with the `err.stack` info?

```bash
${Plugin/Loader} Invalid Options

(${line}:${column} /* from err.stack */) options['foo'] should be string

absolute/path/to/wepback.config.js // from err.stack
1 |  plugins: [
2 |    new FooPlugin({

> 3 |   foo: 1
        ^
4 |    })
```

**Comment by gribnoysup** [comment]
@michael-ciniawsky I pushed my attempt at solving this to this branch. With this approach error message will look like this:

image

There is no wrapper function because I decided to split stack on `webpack.config` to also remove `Module` module trace. I don't think that this 

Parsing would certainly be a cool feature! I think we probably can do it as an improvement separately 😸

File: bin/webpack.js [hunk 13]
```diff
@@ -168,7 +170,23 @@ yargs.parse(process.argv.slice(2), (err, argv, output) => {
 		argv["display"] = "verbose";
 	}
 
-	var options = require("./convert-argv")(yargs, argv);
+	try {
+		Error.stackTraceLimit = 30;
+		var options = require("./convert-argv")(yargs, argv);
+	} catch(err) {
+		var stack = errorHelpers.cutOffAfterFlag(err.stack, "webpack.config");
```
Comment (sokra) [inline bin/webpack.js]: The user could provide any filename for the webpack.config.js via `--config asdf.js`

File: bin/webpack.js [hunk 14]
```diff
@@ -6,6 +6,8 @@
 */
 var path = require("path");
 
+var errorHelpers = require("../lib/ErrorHelpers");
```
Comment (sokra) [inline bin/webpack.js]: `errorHelpers` -> `ErrorHelpers`

File: bin/webpack.js [hunk 15]
```diff
@@ -168,7 +170,23 @@ yargs.parse(process.argv.slice(2), (err, argv, output) => {
 		argv["display"] = "verbose";
 	}
 
-	var options = require("./convert-argv")(yargs, argv);
+	try {
+		Error.stackTraceLimit = 30;
```
Comment (sokra) [inline bin/webpack.js]: Not sure if it's a good idea to set this global variable here.

File: bin/webpack.js [hunk 16]
```diff
@@ -168,7 +170,23 @@ yargs.parse(process.argv.slice(2), (err, argv, output) => {
...
+		var options = require("./convert-argv")(yargs, argv);
+	} catch(err) {
+		var stack = errorHelpers.cutOffAfterFlag(err.stack, "webpack.config");
+		var message = err.message + "\n" + errorHelpers.cutOffMultilineMessage(stack, err.message);
+
+		if(argv.color)
```
Comment (sokra) [inline bin/webpack.js]: add `{` + `}` for better readability

File: bin/webpack.js [hunk 17]
```diff
@@ -168,7 +170,23 @@ yargs.parse(process.argv.slice(2), (err, argv, output) => {
 		argv["display"] = "verbose";
 	}
 
-	var options = require("./convert-argv")(yargs, argv);
+	try {
+		Error.stackTraceLimit = 30;
+		var options = require("./convert-argv")(yargs, argv);
+	} catch(err) {
```
Comment (sokra) [inline bin/webpack.js]: This catches all errors. "Normal" errors from configuration should not be affected by this code. Check error.

File: bin/webpack.js (hunk 15 above)
Comment (gribnoysup) [reply bin/webpack.js]: @sokra the same thing is happening near webpack compiler call lower in the file: 

https://github.com/webpack/webpack/blob/801a25392d99a8ee7f4371c88d7758169b1ac16f/bin/webpack.js#L331

Maybe I can remove both of them and move it to the top of the file?

File: bin/webpack.js (hunk 13 above)
Comment (gribnoysup) [reply bin/webpack.js]: I'll go with the marker function approach then, for now, it will include some traces from `Module` module, but will work with any file name

**Comment by gribnoysup** [comment]
@michael-ciniawsky during the last changes I noticed that not only we need to remove `process.exit ` from `schema-utils`, but `Error.captureStackTrace` also. Right now it erases all stack traces from `ValidationError`

**Comment by michael-ciniawsky** [comment]
### `schema-utils`

- [#17 (no `process.exit(1)`)](https://github.com/webpack-contrib/schema-utils/pull/17)
- [#16 (remove `Error.captureStackTrace`)](https://github.com/webpack-contrib/schema-utils/pull/16)

The necessary changes are merge ready as soon as there aren't any objections left against the current approach taken in webpack core by @gribnoysup 

If we want to implement better `err.message`'s aswell before this lands, please let me know and a quick issue with suggestions/to triage would be 👍. This is very easy to implement, but needs consensus/input about the particular message contents. Otherwise this can also be a follow up 😛 

Sry for the delay I played around if getting an `err.frame` (Code Frame) from the `err.stack` would be an DX improvement, but let's forget about that... 🙃 😛

**Comment by sokra** [comment]
I merge this version into the next branch.

I still like to see the changes in schema-utils. -> separate PR

I also like to see merging the nice error messages from webpack options validation into schema-utils + migration webpack options validation to schema-utils. -> separate PR

**Comment by sokra** [comment]
Thanks

**Comment by gribnoysup** [comment]
@michael-ciniawsky @sokra Yay! Thank you all for cooperation! It was my first PR to webpack and it was a lot of fun! ^.^
